### PR TITLE
WIP. Offline hover preview.

### PR DESCRIPTION
**Pull request by James-Yu**
This is a work in progress. The only TODO is to handle the conflict between `mathjax-node` and `webpack`. In debug mode, things should work. I'm just too tired to go on today. Leaving it to following days.

### CODE REVIEW

**Comment by James-Yu** [comment]
@tamuratak We might join force on fine tuning the preview things. I will handle the webpack.

**Comment by NavyAdmiral** [comment]
What exactly is the problem with compiling mathjax-node with webpack? I tried to compile it on my machine and this was the output:

image


This might be relevant too: https://github.com/mathjax/MathJax-node/issues/332

**Comment by James-Yu** [comment]
@NavyAdmiral It won't load due to the dirty implementation https://github.com/mathjax/MathJax-node/blob/master/lib/main.js#L84 . MathJax-v3 is not being considered due to other reasons. I am thinking of disabling webpack bundling if the problem cannot be solved. All in all the extension activation is not causing great trouble.

**Comment by NavyAdmiral** [comment]
@James-Yu Whats the problem of having online mathjax implementation?

**Comment by James-Yu** [comment]
The key point is that the implementation allows hover preview without webview, which is a great improvement. Offline is a bonus.

**Comment by NavyAdmiral** [comment]
https://github.com/mathjax/mathjax-v3/releases/tag/3.0.0-beta.2 Can this be used instead of mathjax-node?

**Comment by James-Yu** [comment]
Almost not as indicated in https://github.com/mathjax/mj3-demos-node . Hard to translate TeX to SVG.

**Comment by NavyAdmiral** [comment]
According to the creator of mj-node, he's going to deprecate the package when mj3 lands, ref https://github.com/mathjax/MathJax-node/issues/375

**Comment by James-Yu** [comment]
The information is not helping resolve the problem.

**Comment by James-Yu** [comment]
MathJax v3:
```
The current TeX input processor has all the core functionality of the MathJax v2 TeX input, and several of the extensions built in, but some extensions are still to come. For example, \unicode, \bbox, and the color extension are not yet available.

The CommonHTML and SVG output implement all the MathML elements that they do in v2, but do not yet include support for line breaking (neither automatic nor explicit ones); this will be implemented in a later beta version. Both output renderers currently only support the MathJax TeX font; other fonts will be added in the future.
```
These are among the critical reasons why it cannot be used for now.

**Comment by James-Yu** [comment]
It's a hard decision to decide scrap webpack in this branch for NOW. I will go on looking into ways of resolving the issue.
Here is a demo vsix of this branch up to now for testing.
[latex-workshop-5.12.2.vsix.zip](https://github.com/James-Yu/LaTeX-Workshop/files/2485398/latex-workshop-5.12.2.vsix.zip)

**Comment by NavyAdmiral** [comment]
One quick thing I noticed that I think @tamuratak fixed earlier is `undefined custom sequence` 
Try defining:
```
\newcommand*{\R}{\ensuremath{\mathds{R}}}
```

Then, try to preview `$\R \sin{x}...$`

**Comment by NavyAdmiral** [comment]
Since mj-node is not updated too frequently, why don't we make our own fork with a patch for https://github.com/mathjax/MathJax-node/blob/master/lib/main.js#L84 ?

**Comment by James-Yu** [comment]
That is desired. However, its not as simple as locate the correct file location. Trust me, I've tried that, and more problems arise.

**Comment by James-Yu** [comment]
Is there a way for webpack to exclude `node_modules` or more desirably selected modules in the folder? I tried google a solution, but nothing worked.

**Comment by NavyAdmiral** [comment]
@James-Yu Did you take a look at https://www.npmjs.com/package/webpack-node-externals ?

**Comment by James-Yu** [comment]
Because I didn't know about `externals` 🤣 It seems that this one works. Will do some testing and push commits.

**Comment by NavyAdmiral** [comment]
One thing to note, there's no horizontal scrollbar when the svg exceeds hover width.

**Comment by James-Yu** [comment]
Long svg bug see here #876

**Comment by NavyAdmiral** [comment]
That solution is really hacky and not working for even longer SVGs. 
What I'm looking for is the horizontal scrollbar that vscode randomly(?) provides on some hovers.

**Comment by James-Yu** [comment]
Here's what I got. You may notice the scroll.
image

**Comment by NavyAdmiral** [comment]
@James-Yu I edited my comment. That is exactly what I'm looking for, but try reloading vscode and hovering over it for the first time. You will notice that horizontall scrollbar isn't there on the first hover. Is there any reason why this would behave like this?

**Comment by James-Yu** [comment]
I also noticed your observation. I suspect this would be an upstream issue of vscode. All in all there is no difference from the extension side on how mathjax was called. Actually I inspected the generated svg xml and base64 encoded strings, they are identical for the first run and subsequent ones.

**Comment by tamuratak** [comment]
Hi, should I make a PR of my branch to offline-mathjax, or to master?

**Comment by NavyAdmiral** [comment]
There's not too much that can be done on our end until vscode makes max-width of tooltips configurable. This PR looks good to me. @tamuratak has come up with a good solution for colors for now too.

**Comment by James-Yu** [comment]
@tamuratak This is the said color-reading commit. You may directly contribute to this PR, or make PR to this PR/master. We will see which one is better for the feature.

**Comment by tamuratak** [comment]
I prefer my PR. The followings are reasons.
 
1. On Default Light+ theme, this PR does not work well. To parse JSON files for VSCode themes, we support "include" properly, which is not in JSON standard. [link](https://github.com/Microsoft/vscode/blob/master/extensions/theme-defaults/themes/light_plus.json).
```
{
        "$schema": "vscode://schemas/color-theme",
        "name": "Light+ (default light)",
        "include": "./light_vs.json",
        "tokenColors": [
                {
```
2. Default Light+ theme does not define `activityBar.foreground`.  Which color each theme defines depends on each author. One theme defines `editor.foreground`, but another one doesn't. What we really need is [window.getComputedStyle](https://developer.mozilla.org/en-US/docs/Web/API/Window/getComputedStyle) although VSCode doesn't allow it.
3. An appropriate color for math preview in hover depends on each theme. For example, on Ayu Dark theme, this PR causes no error, but math preview is almost invisible. So, for [Aryu](https://marketplace.visualstudio.com/items?itemName=teabyii.ayu) Dark, we must choose a color different from `activityBar.foreground`.


In short, because VSCode themes are in chaos, we had better avoid to touch them directly. This is the reason that I prefer my PR.

Regards,

**Comment by James-Yu** [comment]
@tamuratak Please have a test on this new commit. The W3C way of calculating BW color is used based on the background color of editor, which "should" be defined in any serious themes. Fallbacks to 'type === 'dark', then to black.

**Comment by tamuratak** [comment]
Great. This works really well. I close my PR.

**Comment by tamuratak** [comment]
Please don't forget `rm -rf viewer/mathjax/`.
